models/weight_recorder.cpp

```diff
@@ -0,0 +1,170 @@
+      true,
+      true )
+  , user_set_precise_times_( false )
+  , has_proxies_( false )
+  , local_receiver_( true )
+{
+}
+
+nest::weight_recorder::weight_recorder( const weight_recorder& n )
+  : Node( n )
+  , device_( *this, n.device_ )
+  , user_set_precise_times_( n.user_set_precise_times_ )
+  , has_proxies_( false )
+  , local_receiver_( true )
+{
+}
+
+void
+nest::weight_recorder::init_state_( const Node& np )
+{
+  const weight_recorder& wr = dynamic_cast< const weight_recorder& >( np );
+  device_.init_state( wr.device_ );
+  init_buffers_();
+}
+
+void
+nest::weight_recorder::init_buffers_()
+{
+  device_.init_buffers();
+  B_.events_ = std::vector< WeightRecorderEvent >();
+}
+
+void
+nest::weight_recorder::calibrate()
+{
+  if ( kernel().event_delivery_manager.get_off_grid_communication()
+    and not device_.is_precise_times_user_set() )
+  {
+    device_.set_precise_times( true );
+    std::string msg = String::compose(
+      "Precise neuron models exist: the property precise_times "
+      "of the %1 with gid %2 has been set to true",
+      get_name(),
+      get_gid() );
+
+    if ( device_.is_precision_user_set() )
+    {
+      // if user explicitly set the precision, there is no need to do anything.
+      msg += ".";
+    }
+
+    else
+    {
+      // it makes sense to increase the precision if precise models are used.
+      device_.set_precision( 15 );
+      msg += ", precision has been set to 15.";
+    }
+
+    LOG( M_INFO, "spike_detector::calibrate", msg );
+  }
+
+  device_.calibrate();
+}
+
+void
+nest::weight_recorder::update( Time const&, const long from, const long to )
+{
+
+  for ( std::vector< WeightRecorderEvent >::iterator e = B_.events_.begin();
+        e != B_.events_.end();
+        ++e )
+  {
+    device_.record_event( *e );
+  }
+
+  // do not use swap here to clear, since we want to keep the reserved()
+  // memory for the next round
+  B_.events_.clear();
+}
+
+void
+nest::weight_recorder::get_status( DictionaryDatum& d ) const
+{
+  // get the data from the device
+  device_.get_status( d );
+
+  // if we are the device on thread 0, also get the data from the
+  // siblings on other threads
+  if ( local_receiver_ && get_thread() == 0 )
+  {
+    const SiblingContainer* siblings =
+      kernel().node_manager.get_thread_siblings( get_gid() );
+    std::vector< Node* >::const_iterator sibling;
+    for ( sibling = siblings->begin() + 1; sibling != siblings->end();
+          ++sibling )
+      ( *sibling )->get_status( d );
+  }
+}
+
+void
+nest::weight_recorder::set_status( const DictionaryDatum& d )
+{
+  if ( d->known( names::precise_times ) )
+    user_set_precise_times_ = true;
+
+  device_.set_status( d );
+}
+
+void
+nest::weight_recorder::handle( WeightRecorderEvent& e )
+{
+  // accept spikes only if detector was active when spike was
+  // emitted
+  if ( device_.is_active( e.get_stamp() ) )
+  {
+    WeightRecorderEvent* event = e.clone();
+    B_.events_.push_back( *event );
```

Comment:
It might be better to check for sender/receiver here, to avoid events_ getting swamped in a large simulation. I suggest testing for performance in a large network.